### PR TITLE
Fix java and c# bindings + cleanup

### PR DESCRIPTION
**Pull request by drdanz**
This fixes C# and java bindings for me... Commit 0750722 is critical, @traversaro please review it, since I don't know what I'm doing 😆 

Tested (locally) build for
* C#
* Java
* Lua
* Python
* Octave
* TCL/TK
* Ruby
* Perl

They all build, but I haven't actually tried using them

### CODE REVIEW

**Comment by traversaro** [comment]
Unfortunately this PR is breaking Lua support : 
~~~
traversaro@turati:~/src/isaac-superbuild/isaac/YARP/bindings$ lua ./tests/lua/test_string.lua 
lua: ./tests/lua/test_string.lua:11: Wrong arguments for overloaded function 'Contactable_open'
  Possible C/C++ prototypes are:
    yarp::os::Contactable::open()
    yarp::os::Contactable::open(yarp::os::ConstString const &)
    yarp::os::Contactable::open(yarp::os::Contact const &,bool)
    yarp::os::Contactable::open(yarp::os::Contact const &)

stack traceback:
	[C]: in function 'open'
	./tests/lua/test_string.lua:11: in main chunk
	[C]: ?
~~~

File: bindings/tests/lua/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+# test_string.lua uses the yarpserver, so for now it is not include in the CTest infrastructure
+
+macro(add_lua_unit_test luascript)
+    # Find the lua interpreter (ideally, this should be handled in FindLua or in a separate
+    # FindLuaInterp like for python)
+    find_program(LUA_EXECUTABLE lua)
+    if(LUA_EXECUTABLE)
+        add_test(NAME ${luascript} COMMAND ${LUA_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/${luascript})
+        set_tests_properties(${luascript} PROPERTIES ENVIRONMENT "LUA_CPATH=\;\;\;$<TARGET_FILE:${target_name}>")
```
Comment (drdanz) [inline bindings/tests/lua/CMakeLists.txt]: ❤️

**Comment by drdanz** [comment]
rebased

File: bindings/tests/lua/CMakeLists.txt (hunk 1 above)
Comment (traversaro) [reply bindings/tests/lua/CMakeLists.txt]: Thanks to the heart I just realized that this line will fail on CMake < 3.3 , so I should add an if for that case!

**Comment by traversaro** [comment]
Done in https://github.com/robotology/yarp/pull/1350/commits/d79afdf8072cfafdc05406203acd4f430cc60ae6 .

**Comment by traversaro** [comment]
Unfortunately, the requirement of CMake 3.3 means that the test will not run in the Travis Linux build, unless we install the `cmake3` Trusty package: https://packages.ubuntu.com/trusty/cmake3 .

**Comment by drdanz** [comment]
We can easily install the cmake3, I'm not sure how it works on trusty though... Does it replace the "cmake" package or should we call "cmake3" instead?

**Comment by drdanz** [comment]
Actually I tested and it *replaces* the cmake package so we can just install it...

**Comment by drdanz** [comment]
Let's see if it builds with cmake3

**Comment by traversaro** [comment]
To be honest I was going to propose this, but I am bit afraid that this will leave us expose to problems related to "old" CMake version (as the one shipped by default in the Debian that we support), given that in the CI we would test CMake 3.5 (Trusty), and always the latest CMake for macOs/Windows.

**Comment by drdanz** [comment]
I was confirmed that iCub stretch images are almost ready, so I think we can just forget about CMake 3.0 on travis.
Also CMake 3.6 is available on debian jessie backports and the 3 months support deadline is already passed, so it should be ok to switch to CMake 3.5 as soon as we have the stretch images...

**Comment by francesco-romano** [comment]
```
/Users/makaveli/Projects/src/yarp/build/generated_src/java/yarpJAVA_wrap.cxx:14162:28: error: too few arguments to function call, expected 2, have 1
  (&arg2)->assign(arg2_pstr);
  ~~~~~~~~~~~~~~~          ^
In file included from /Users/makaveli/Projects/src/yarp/build/generated_src/java/yarpJAVA_wrap.cxx:696:
In file included from /Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/all.h:19:
In file included from /Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/BinPortable.h:10:
In file included from /Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/Portable.h:10:
In file included from /Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/PortReader.h:10:
In file included from /Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/ConnectionReader.h:10:
/Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/ConstString.h:310:5: note: 'assign' declared here
    inline ConstString& assign(const char *s, size_t n) {
    ^
```

**Comment by francesco-romano** [comment]
```
/Users/makaveli/Projects/src/yarp/build/generated_src/java/yarpJAVA_wrap.cxx:14158:28: error: too few arguments to function call, expected 2, have 1
  (&arg2)->assign(arg2_pstr);
  ~~~~~~~~~~~~~~~          ^
In file included from /Users/makaveli/Projects/src/yarp/build/generated_src/java/yarpJAVA_wrap.cxx:692:
In file included from /Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/all.h:19:
In file included from /Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/BinPortable.h:10:
In file included from /Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/Portable.h:10:
In file included from /Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/PortReader.h:10:
In file included from /Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/ConnectionReader.h:10:
/Users/makaveli/Projects/src/yarp/src/libYARP_OS/include/yarp/os/ConstString.h:310:5: note: 'assign' declared here
    inline ConstString& assign(const char *s, size_t n) {
    ^
```

**Pull request merged**